### ISSUE TITLE
Smoke Test Brief — Z-Curvature Features for Genomic Tasks

### ISSUE THREAD

**Issue by zfifteen** [zfifteen/unified-framework#503]
⸻

🚦 Smoke Test Brief — Z-Curvature Features for Genomic Tasks

Background (why this exists)

We hypothesize that simple Z-curvature features—derived from a φ-phase index and \theta’(i,k)=\varphi\cdot\{i/\varphi\}^k—capture “prime-like” sparsity/structure that may add predictive signal on genomic tasks (mutation hotspots, CRISPR off-targets).
This smoke test is not about state-of-the-art; it’s a cheap, same-day check: do tiny Z-features move the needle at all versus a minimal baseline? If yes → greenlight a proper study. If no → fail fast.

Purpose (what the smoke test must answer)
	1.	On a tiny, well-controlled dataset, do Z-features provide any measurable lift over a naive sequence baseline?
	2.	Is that lift robust to trivial changes (window size and k)?
	3.	Is the compute cost negligible (can run in <1–2 minutes on a laptop)?

⸻

Scope (keep it tiny)
	•	Tasks (pick one first):
	•	A) Mutation hotspot classification (binary): label windows as mutated (1) vs neutral (0).
	•	B) CRISPR off-target scoring (binary): known off-targets (1) vs random decoys (0).
	•	Data (fixtures, minimal):
	•	Create two CSVs with ≤1–2k rows each.
	•	mutation_windows.csv: columns chrom,start,end,seq,label
	•	crispr_pairs.csv: columns guide,site,seq,label (site=genomic protospacer ± PAM; label=1 if validated off-target)
	•	It’s OK to hand-craft fixtures or sample from known examples; this is a sanity check, not a paper.

⸻

Deliverables (files Copilot must produce)

smoke_test/
  README.md                         # how to run, what to expect
  requirements.txt                  # pin minimal deps
  run_smoke_test.py                 # CLI runner: trains/evaluates baselines vs Z
  z_features.py                     # φ-phase & θ' feature extraction (tiny)
  seq_features.py                   # minimal baseline features
  data/
    mutation_windows.csv            # tiny fixture (≤1–2k rows)
    crispr_pairs.csv                # tiny fixture (≤1–2k rows)
  outputs/
    metrics.json                    # PR-AUC, AUROC, top-k recall where applicable
    feature_importance.csv          # model-driven ranking
    ablation_table.csv              # baseline vs +Z across settings

⸻

Minimal feature sets

Baseline (keep dumb on purpose)
	•	gc = GC fraction in the window
	•	kmer3_entropy = entropy of 3-mer distribution (or sum of top-N 3-mers)
	•	homopolymer_max = length of longest homopolymer
	•	(CRISPR only) mismatch_count between guide and site (Hamming on 20-nt)

Z-curvature (your hypothesis)
	•	phi_phase_mean = mean of \{\tfrac{i}{\varphi}\} over indices in the window
	•	phi_phase_std
	•	theta_mean(k) = mean of \theta’(i,k)
	•	theta_std(k)
	•	theta_energy_3bp(k) = (optional) product of 3-bp periodicity amplitude and theta_mean(k) (single scalar)

Hyper-params to test: k \in \{0.1, 0.3, 0.04449\} and window size w \in \{51, 101\}.
Set \varphi = (1+\sqrt{5})/2. Use 0-based index for i.

⸻

Modeling & Metrics
	•	Model: Logistic Regression (liblinear) and XGBoost (tiny depth). Pick best of the two per setting.
	•	Splits: Stratified 80/20 with fixed seed 42. No leakage.
	•	Metrics:
	•	Primary: PR-AUC (positives are rare).
	•	Secondary: AUROC.
	•	(CRISPR only) Recall@K (K = 50) on the ranked list.

⸻

Acceptance criteria (go/no-go)
	•	PASS if any Z-augmented setting beats baseline-only by:
	•	ΔPR-AUC ≥ +0.02 absolute, and
	•	effect is consistent across at least 2 of the 4 combos:
(k in {0.1, 0.3, 0.04449}) × (w in {51, 101})
	•	FAIL if Δ < 0.02 everywhere or improvements flip sign across trivial changes.
	•	Compute: end-to-end runtime ≤ 2 minutes on a typical laptop.

⸻

Implementation checklist (what to code)

1) z_features.py
	•	phi = (1 + 5**0.5)/2
	•	Given a sequence window, iterate index i = 0..w-1:
	•	phi_phase_i = frac(i / phi)  (use x - floor(x))
	•	theta_i = phi * (phi_phase_i ** k)
	•	Return dict of: phi_phase_mean, phi_phase_std, theta_mean_k{K}, theta_std_k{K} (format the names with K).

2) seq_features.py
	•	Implement gc, kmer3_entropy, homopolymer_max.
	•	(CRISPR) add mismatch_count(guide,site).

3) run_smoke_test.py (CLI)
	•	Args: --task [mutation|crispr], --window {51,101}, --k {0.1,0.3,0.04449}, --model [logreg|xgb]
	•	Pipeline:
	1.	Load CSV.
	2.	Build features: Baseline → Baseline+Z.
	3.	Train/val split (seed=42).
	4.	Fit model; score val set.
	5.	Write metrics.json, feature_importance.csv.
	•	Also run the grid: all (k, window) × both models; aggregate into ablation_table.csv.

4) README.md (usage)
	•	Install: python -m venv .venv && source .venv/bin/activate && pip install -r requirements.txt
	•	Run single setting:
	•	python smoke_test/run_smoke_test.py --task mutation --window 101 --k 0.3 --model xgb
	•	Run full grid (include a tiny shell snippet).
	•	How to read outputs/metrics.json and the acceptance criteria.

5) requirements.txt

biopython==1.83
numpy==2.0.1
pandas==2.2.2
scikit-learn==1.5.1
xgboost==2.1.1

⸻

Pseudocode (for Copilot to follow)

# run_smoke_test.py (sketch)
args = parse_args()
df = pd.read_csv(data_path_for(args.task))
X_base, y = build_baseline_features(df, window=args.window, task=args.task)
X_z = add_z_features(X_base.copy(), window=args.window, k=args.k, task=args.task)

for feature_set_name, X in [("baseline", X_base), ("baseline_plus_z", X_z)]:
    X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2,
                                                      stratify=y, random_state=42)
    model = get_model(args.model)
    model.fit(X_train, y_train)
    p = model.predict_proba(X_val)[:,1]
    pr_auc = average_precision_score(y_val, p)
    auroc = roc_auc_score(y_val, p)
    save_metrics(feature_set_name, pr_auc, auroc, extras_for_task(args.task, df, p, y_val))
save_ablation_table()
save_feature_importance(model, X.columns)

⸻

Guardrails (keep Copilot honest)
	•	No deep learning, no external datasets, no internet calls.
	•	Keep window extraction and indexing deterministic.
	•	If a feature can’t be computed (e.g., short seq), drop row and log count.
	•	Always report class balance and number of valid samples per run.
	•	Fix random_state=42 everywhere; log start/end time.

⸻

Expected outputs (example)
	•	outputs/metrics.json:

{
  "task": "mutation",
  "window": 101,
  "k": 0.3,
  "model": "xgb",
  "baseline": {"pr_auc": 0.312, "auroc": 0.664, "n_val": 200},
  "baseline_plus_z": {"pr_auc": 0.339, "auroc": 0.683, "n_val": 200},
  "delta": {"pr_auc": 0.027, "auroc": 0.019}
}

	•	outputs/ablation_table.csv: rows for each (k, w, model) with baseline, +Z, and deltas.
	•	outputs/feature_importance.csv: top features; verify Z features appear when Δ>0.

⸻

Decision rule (what I will read first)
	1.	ablation_table.csv: Are there ≥2 settings with ΔPR-AUC ≥ +0.02?
	2.	feature_importance.csv: Do Z features show non-trivial importance in those wins?
	3.	Runtime under 2 min?

If (1)&(2)&(3) → Greenlight deeper build (bigger data, DNA shape, RT/nucleosomes). Else → Stop.
